File: doc-examples/hibernate-example-java/src/main/java/example/ProductManager.java

package example;

import io.micronaut.transaction.TransactionOperations;
import jakarta.inject.Singleton;
import jakarta.persistence.EntityManager;
import org.hibernate.Session;

@Singleton
public class ProductManager {

    private final EntityManager entityManager;
    private final TransactionOperations<Session> transactionManager;

    public ProductManager(EntityManager entityManager,
                          TransactionOperations<Session> transactionManager) { // <1>
        this.entityManager = entityManager;
        this.transactionManager = transactionManager;
    }

    Product save(String name, Manufacturer manufacturer) {
        return transactionManager.executeWrite(status -> { // <2>
            final Product product = new Product(name, manufacturer);
            entityManager.persist(product);
            return product;
        });
    }

    Product find(String name) {
        return transactionManager.executeRead(status -> // <3>
            status.getConnection().createQuery("from Product p where p.name = :name", Product.class)
                .setParameter("name", name)
                .getSingleResult()
        );
    }
}

File: doc-examples/hibernate-example-python/src/main/python/example/ProductManager.py
from jakarta.inject import Singleton
from jakarta.persistence import EntityManager
from micronaut.transaction import TransactionOperations
from org.hibernate import Session

from example.Manufacturer import Manufacturer
from example.Product import Product


@Singleton
class ProductManager:

    def __init__(self, entity_manager: EntityManager, transaction_manager: TransactionOperations[Session]):  # <1>
        self.entity_manager = entity_manager
        self.transaction_manager = transaction_manager

    def save(self, name: str, manufacturer: Manufacturer) -> Product:
        def persist(status):  # <2>
            product = Product(name, manufacturer)
            self.entity_manager.persist(product)
            return product
        return self.transaction_manager.executeWrite(persist)

    def find(self, name: str) -> Product:
        return self.transaction_manager.executeRead(  # <3>
            lambda status: status.getConnection().createQuery("from Product p where p.name = :name", Product)
            .setParameter("name", name)
            .getSingleResult()
        )
